### PR TITLE
4601 display dollar currency price

### PR DESCRIPTION
**Pull request by andrewkryshtal**

### CODE REVIEW

**Review by keshan3262 — APPROVED** [review]

File: src/components/activity-groups-list/activity-group-item/activity-group-amount-change/activity-group-amount-change.tsx [hunk 1]
```diff
@@ -16,35 +22,59 @@ interface Props {
...
+
+      amounts.push({
+        parsedAmount,
+        isPositive,
+        symbol,
+        exchangeRate
+      });
+    }
+
+    const dollarSum = formatAssetAmount(new BigNumber(summaryPositiveAmount), BigNumber.ROUND_DOWN, 2);
```
Comment (unixvb) [inline src/components/activity-groups-list/activity-group-item/activity-group-amount-change/activity-group-amount-change.tsx:61]: use the same naming for `summaryPositiveAmount` => `positiveAmountSum`

File: src/store/currency/currency-state.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+export interface TokenExchangeRate {
+  tokenAddress: string;
+  tokenId?: number;
+  exchangeRate: string;
+}
+
+export interface MappedTokenExchangeRate {
+  [key: string]: number;
+}
```
Comment (unixvb) [inline src/store/currency/currency-state.ts:12]: use `Record`

File: src/store/currency/currency-state.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { createEntity } from '../create-entity';
+import { LoadableEntityState } from '../types';
+
+export interface TokenExchangeRate {
+  tokenAddress: string;
+  tokenId?: number;
+  exchangeRate: string;
+}
```
Comment (unixvb) [inline src/store/currency/currency-state.ts:8]: move this to a separate file.  
This type is not related to the state.

File: src/store/currency/currency-state.ts (hunk 2 above)
Comment (unixvb) [reply src/store/currency/currency-state.ts:12]: and rename to `TokenExchangeRateRecord`

File: src/store/currency/currency-state.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+export interface MappedTokenExchangeRate {
+  [key: string]: number;
+}
+
+export interface exchangeRatesState {
```
Comment (unixvb) [inline src/store/currency/currency-state.ts:14]: interface names should start with the cappital `ExchangeRatesState`

File: src/token/utils/token.utils.ts [hunk 5]
```diff
@@ -3,6 +3,11 @@ import { TokenTypeEnum } from '../../interfaces/token-type.enum';
 import { TokenMethodsAssertionsMap } from '../data/token-methods-assertions';
 
 export const tokenMetadataSlug = <T extends { address: string; id?: number }>({ address, id }: T) => `${address}_${id}`;
+export const getTokenAddressFromSlug = (slug: string): { tokenAddress: string; tokenId: string } => {
+  const splittedString = slug.split('_');
+
+  return { tokenAddress: splittedString[0], tokenId: splittedString[1] };
```
Comment (unixvb) [inline src/token/utils/token.utils.ts:9]: just return `{ address, id }` we have `Token` in name of fn already

File: src/utils/number.util.ts [hunk 6]
```diff
@@ -1,4 +1,16 @@
...
+import { isDefined } from './is-defined';
+
+export const formatAssetAmount = (
+  amount: BigNumber,
+  roundingMode: BigNumber.RoundingMode = BigNumber.ROUND_DOWN,
+  decimalPlace: number | undefined = undefined
+) => {
+  return amount
+    .decimalPlaces(
+      amount.abs().lt(1000) && !isDefined(decimalPlace) ? 6 : isDefined(decimalPlace) ? decimalPlace : 2,
```
Comment (unixvb) [inline src/utils/number.util.ts:12]: refactor this, please

File: src/store/currency/currency-reducers.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    ...state,
+    tokensExchangeRates: createEntity({}, true)
+  }));
+  builder.addCase(loadExchangeRates.success, (state, { payload }) => ({
+    ...state,
+    tokensExchangeRates: createEntity(payload)
+  }));
+  builder.addCase(loadExchangeRates.fail, (state, { payload }) => ({
+    ...state,
+    tokensExchangeRates: createEntity(state.tokensExchangeRates.data || null, false, payload)
```
Comment (unixvb) [inline src/store/currency/currency-reducers.ts:18]: why `state.tokensExchangeRates.data || null`?

File: src/store/currency/currency-epics.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+export const loadExchangeRatesEpic = (action$: Observable<Action>) =>
+  action$.pipe(
+    ofType(loadExchangeRates.submit),
+    switchMap(() =>
+      from(templeWalletApi.get<TokenExchangeRate[]>('exchange-rates')).pipe(
+        map(({ data }) => {
+          const mappedRates = Object.fromEntries(
+            data.map(({ tokenAddress, exchangeRate }) => [tokenAddress, Number(exchangeRate)])
+          );
```
Comment (unixvb) [inline src/store/currency/currency-epics.ts:19]: optimize this algorithm. 
use single loop through `data` instead of 2.

File: src/store/currency/currency-epics.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  );
+
+export const loadTezosExchangeRateEpic = (action$: Observable<Action>) =>
+  action$.pipe(
+    ofType(loadTezosExchangeRate.submit),
+    switchMap(() =>
+      from(templeWalletApi.get<number>('exchange-rates/tez')).pipe(
+        map(({ data }) => {
+          return loadTezosExchangeRate.success(data);
+        }),
```
Comment (unixvb) [inline src/store/currency/currency-epics.ts:35]: make short return here

File: src/store/currency/currency-reducers.ts (hunk 7 above)
Comment (andrewkryshtal) [reply src/store/currency/currency-reducers.ts:18]: because first load could be rejected, and there could be no data saved before

File: src/navigator/main-stack.tsx [hunk 10]
```diff
@@ -49,6 +51,11 @@ export const MainStackScreen = () => {
 
   useAppLockTimer();
   useBeaconHandler();
+  useEffect(() => {
+    dispatch(loadExchangeRates.submit());
+    dispatch(loadTezosExchangeRate.submit());
+  }, []);
```
Comment (unixvb) [inline src/navigator/main-stack.tsx:57]: do we really need this?  
I think `useEffect` with timer will do the job

File: src/screens/wallet/token-list/token-list.tsx [hunk 11]
```diff
@@ -61,15 +63,17 @@ export const TokenList: FC = () => {
...
             )}
 
             {filteredTokensList.map(
-              token =>
+              (token, index) =>
                 token.isVisible && (
                   <TokenListItem
-                    key={token.address}
+                    exchangeRate={tokensExchangeRates.data[token.address]}
+                    key={token.address + index}
```
Comment (unixvb) [inline src/screens/wallet/token-list/token-list.tsx:76]: key should be the first prop

File: src/store/currency/currency-reducers.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
+import { createReducer } from '@reduxjs/toolkit';
+
+import { createEntity } from '../create-entity';
+import { loadExchangeRates, loadTezosExchangeRate } from './currency-actions';
+import { currencyInitialState, ExchangeRatesState } from './currency-state';
+
+export const currencyReducers = createReducer<ExchangeRatesState>(currencyInitialState, builder => {
+  builder.addCase(loadExchangeRates.submit, state => ({
+    ...state,
+    tokensExchangeRates: createEntity({}, true)
```
Comment (unixvb) [inline src/store/currency/currency-reducers.ts:10]: you are updating value on "`submit`" action, this will cause blinking of $ values everywhere

File: src/store/currency/currency-reducers.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    ...state,
+    tokensExchangeRates: createEntity(payload)
+  }));
+  builder.addCase(loadExchangeRates.fail, (state, { payload }) => ({
+    ...state,
+    tokensExchangeRates: createEntity(state.tokensExchangeRates.data, false, payload)
+  }));
+  builder.addCase(loadTezosExchangeRate.submit, state => ({
+    ...state,
+    tezosExchangeRate: createEntity(0, true)
```
Comment (unixvb) [inline src/store/currency/currency-reducers.ts:22]: the same

File: src/components/dollar-equivalent-text/dollar-equivalent-text.tsx [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+  if (isDefined(exchangeRate)) {
+    formattedDollarEquivalent = formatAssetAmount(
+      new BigNumber(Number(balance) * exchangeRate),
+      BigNumber.ROUND_DOWN,
+      2
+    );
+  }
+
+  return formattedDollarEquivalent ? <Text style={style}>{formattedDollarEquivalent} $</Text> : null;
+};
```
Comment (unixvb) [inline src/components/dollar-equivalent-text/dollar-equivalent-text.tsx:25]: You could use short return here 
```
const Component = () => condition ?  <Text>{formatAssetAmount(...)}</Text> : null; 
```

File: src/components/dollar-equivalent-text/dollar-equivalent-text.tsx [hunk 15]
```diff
@@ -0,0 +1,25 @@
+import { BigNumber } from 'bignumber.js';
+import React, { FC } from 'react';
+import { StyleProp, TextStyle, Text } from 'react-native';
+
+import { isDefined } from '../../utils/is-defined';
+import { formatAssetAmount } from '../../utils/number.util';
+
+interface Props {
+  balance: string;
+  exchangeRate: number | undefined;
```
Comment (unixvb) [inline src/components/dollar-equivalent-text/dollar-equivalent-text.tsx:10]: `exchangeRate?: number`

File: src/modals/confirmation-modal/operations-confirmation/fee-form-input/fee-form-input.tsx [hunk 16]
```diff
@@ -47,6 +50,18 @@ export const FeeFormInput: FC<Props> = ({
...
+    BigNumber.ROUND_UP,
+    2
+  );
+
+  const formattedStorageFeeInDollarEquivalent = formatAssetAmount(
+    new BigNumber(Number(storageFee) * exchangeRate),
+    BigNumber.ROUND_UP,
+    2
+  );
+
```
Comment (unixvb) [inline src/modals/confirmation-modal/operations-confirmation/fee-form-input/fee-form-input.tsx:64]: could you reuse `DollarEquivalentText` component?

**Review by unixvb — APPROVED** [review]
